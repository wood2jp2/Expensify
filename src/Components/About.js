import React from 'react'

const About = () => <p>ABOUT</p>

export default About