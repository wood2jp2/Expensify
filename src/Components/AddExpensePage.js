import React from 'react'

const AddExpensePage = () => (
    <div>
        Add Expense component!
    </div>
)

export default AddExpensePage