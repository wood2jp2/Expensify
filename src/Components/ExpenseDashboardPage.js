import React from 'react'

const ExpenseDashboardPage = () => (
    <div>
        Dashboard component
    </div>
)

export default ExpenseDashboardPage