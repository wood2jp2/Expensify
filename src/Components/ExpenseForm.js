import React, { Component } from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends Component {

    state = {
        description: this.props.description || '',
        amount: this.props.amount ? (this.props.amount / 100).toString() : '',
        note: this.props.note || '',
        createdAt: moment(this.props.createdAt) || moment(),
        calendarFocused: false,
        error: ''
    }

    onDescriptionChange = e => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onAmountChange = e => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onNoteChange = e => {
        const note = e.target.value
        this.setState(() => ({note}))
    }

    onDateChange = createdAt => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }

    onSubmit = e => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            // Please provide description and amount
            this.setState(() => ({error: 'Please provide a description and amount within the fields.'}))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    onFocusChange = ({focused}) => this.setState(() => ({calendarFocused: focused}))

    render() {
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}

                <form onSubmit={this.onSubmit}>
                {/* autoFocus puts the cursor automatically in that field on load */}
                    <input 
                        autoFocus 
                        placeholder="Description" 
                        type="text"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        placeholder="Amount" 
                        type="text"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        id=""
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)."
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button
                        // disabled={!this.state.description || !this.state.amount}
                    >Add Expense</button>
                </form>
            </div>
        )
    }
}

 