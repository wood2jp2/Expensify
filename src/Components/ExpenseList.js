import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../Selectors/expenses'

export const ExpenseList = props => (
    <div>
    {
        props.expenses.length === 0 ? (
            <p>There are no expenses to display</p>
        ) : (props.expenses && props.expenses.map((expense, i) => 
            <ExpenseListItem 
                key={i}
                {...expense} />)
        )
    }
        
    </div>
)

const mapStateToProps = state => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList