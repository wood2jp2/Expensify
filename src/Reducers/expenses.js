// Expenses Reducer and default values
const expensesReducerDefaultState = []

// The action is the result of the action functions
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [...state, action.expense]
        case "REMOVE_EXPENSE":
            return state.filter(({ id }) => id !== action.id)
        case "EDIT_EXPENSE":
            return state.map(expense => {
                return expense.id === action.id ?
                    {...expense, ...action.updates} :
                    expense
            })
        default: 
            return state 
    }
}