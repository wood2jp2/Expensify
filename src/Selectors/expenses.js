import moment from 'moment'

export default (expensesState, {text, sortBy, startDate, endDate}) => {
    return expensesState.filter(expense => {
        const 
             createdAtMoment = moment(expense.createdAt),
             startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true,
             endDateMatch  = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true,
             textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        // Should be sorted most recent expense to oldest expense
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt 
        } 
        // Should be sorted most expensive to least expensive
        else if (sortBy === 'amount') {
            return b.amount - a.amount
        }
    })
}