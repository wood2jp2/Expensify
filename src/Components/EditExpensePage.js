import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../Actions/expenses'

const EditExpensePage = props => {
    return (
        <div>
            <ExpenseForm 
                { ...props.expense } 
                onSubmit={ expense => {
                    props.dispatch(editExpense(props.expense.id, expense))
                    props.history.push('/')
                }}
            />
            <button onClick={ () => {
                props.dispatch(removeExpense({id: props.expense.id}))
                props.history.push('/')
            }}>Remove</button>
        </div>
    )
}


const mapStateToProps = (state, props) => ({ expense: state.expenses.find( expense => expense.id === props.match.params.id) })

export default connect(mapStateToProps)(EditExpensePage)