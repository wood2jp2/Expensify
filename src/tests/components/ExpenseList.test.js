import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../Components/ExpenseList'
import { expenses } from '../fixtures/expenses'

test("should render the ExpenseList with expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test("should render ExpenseList with empty message", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})