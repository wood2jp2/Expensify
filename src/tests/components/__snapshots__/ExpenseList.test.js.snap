// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    There are no expenses to display
  </p>
</div>
`;

exports[`should render the ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={1950}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={45000}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="2"
    note=""
  />
</div>
`;
